ada_2_fhir-r4/zibs2020/payload/0.0/nl-core-Patient.xsl

```diff
@@ -0,0 +1,360 @@
+
+The full text of the license is available at http://www.gnu.org/copyleft/lesser.html
+-->
+
+<xsl:stylesheet exclude-result-prefixes="#all"
+    xmlns="http://hl7.org/fhir"
+    xmlns:util="urn:hl7:utilities" 
+    xmlns:f="http://hl7.org/fhir" 
+    xmlns:xd="http://www.oxygenxml.com/ns/doc/xsl"
+    xmlns:nf="http://www.nictiz.nl/functions" 
+    xmlns:nm="http://www.nictiz.nl/mappings"
+    xmlns:uuid="http://www.uuid.org"
+    xmlns:xs="http://www.w3.org/2001/XMLSchema" 
+    xmlns:xsl="http://www.w3.org/1999/XSL/Transform" 
+    version="2.0">
+    
+    <xsl:output method="xml" indent="yes"/>
+    <xsl:strip-space elements="*"/>
+    
+    <xsl:variable name="referById" select="false()"/>
+    
+    <xd:doc>
+        <xd:desc>Global variable that can be called anywhere in the process to access, include or reference any unique patient present in the input data</xd:desc>
+    </xd:doc>
+    <!-- Based on variable 'patients' in _zib2017.xsl -->
+    <xsl:template name="getAllPatients" as="element()*">
+        <xsl:param name="in" select="."/>
+        <!-- Originally, this grouping contained the 'nf:ada-pat-id' function, but because this function is not used outside this variable I edited it out. Maybe it should be removed from 2_fhir_fhir_include and added to _zib2017? -->
+        <xsl:variable name="identifier" select="(identificatienummer[@root = $oidBurgerservicenummer],identificatienummer[not(@root = $oidBurgerservicenummer)])[1]"/>
+        <xsl:for-each-group select="$in//patient[not(patient)][not(@datatype = 'reference')][.//(@value | @code | @nullFlavor)]" group-by="concat($identifier/@root, $identifier/@value)">
+            <!-- This for-each-group should be enough to sort patients by their official identifier, but $patients does a second group-by. Apparently there are enough use cases where no identificatienummer is present (or masked?), or dummy identificatienummers that exist in multiple patients. So I guess this second group-by is necessary? If a use case uses the same patient multiple times, it still gets filtered out -->
+            <!-- nf:getGroupingKeyPatient() is a function that concatenates a lot of Patient values together. It is not only used here, but also when generating the reference -->
+            <xsl:for-each-group select="current-group()" group-by="nf:getGroupingKeyPatient(.)">
+                <!-- Why did the logic below exist? -->
+                <!--<!-\- uuid als fullUrl en ook een fhir id genereren vanaf de tweede groep -\->
+                <xsl:variable name="uuid" as="xs:boolean" select="position() > 1"/>-->
+                <!-- Originally an 'xmlns=""' was added to the 'unieke-patient' element to prevent namespace issues. We solve this by placing the construction below in its separate nm namespace. Implemented the proposed <resource type="..."> construction. Maybe we can combine variables like $patient and $healthprofessional and ... in the near future? -->
```

Comment:
Ja, ik denk niet dat het schaalbaar is om voor elke zib een eigen variabele te maken. Mede daarom voel ik er meer voor om een grote variabele met alleen de metadata te maken.

Reply:
De `$uuid` constructie was trouwens zodat de 1e (patient, zorgverlener) een gegenereerd id kreeg, en de tweede daarvan met alleen een iets afwijkende naam of adres niet dezelfde, maar altijd een uuid. Vandaar de `position() > 1`